src/core/lib/transport/metadata.cc

```diff
@@ -223,11 +230,14 @@ void grpc_mdctx_global_shutdown() {
   }
 }
 
+#ifndef NDEBUG
 static int is_mdelem_static(grpc_mdelem e) {
-  return GRPC_MDELEM_DATA(e) >= &grpc_static_mdelem_table[0] &&
-         GRPC_MDELEM_DATA(e) <
+  return reinterpret_cast<grpc_core::StaticMetadata*>(GRPC_MDELEM_DATA(e)) >=
+             &grpc_static_mdelem_table[0] &&
+         reinterpret_cast<grpc_core::StaticMetadata*>(GRPC_MDELEM_DATA(e)) <
```

Comment:
It would indeed be illegal to deref/access it. However, here we are just doing a pointer comparison, which is valid: we're only testing if some pointer "e" is in fact < some pointer addrof(grpc_static_mdelem_table) + some constant.